dev/com.ibm.ws.logging/src/com/ibm/ws/logging/utils/FileLogHolder.java

```diff
@@ -286,17 +287,35 @@ public synchronized void close() {
      * @param record
      */
     @Override
-    public synchronized void writeRecord(String record) {
-        long length = record.length() + LoggingConstants.nlen;
-        PrintStream ps = getPrintStream(length);
-        ps.println(record);
-        if (ps.checkError()) {
-            setStreamStatus(StreamStatus.CLOSED, null, null, DummyOutputStream.psInstance);
-            // to avoid junit test to print an error message
-            if (System.getProperty("test.classesDir") == null && System.getProperty("test.buildDir") == null) {
-                System.err.println(Tr.formatMessage(getTc(), "FAILED_TO_WRITE_LOG", new Object[] { getPrimaryFile().getAbsolutePath() }));
+    public void writeRecord(String record) {
+
+        ThreadLocal<Boolean> isError = ThreadLocal.withInitial(() -> false);
+
+        synchronized (this) {
+            long length = record.length() + LoggingConstants.nlen;
+            PrintStream ps = getPrintStream(length);
+            ps.println(record);
+            if (ps.checkError()) {
+                setStreamStatus(StreamStatus.CLOSED, null, null, DummyOutputStream.psInstance);
+                isError.set(true);
             }
         }
+        /*
+         * To avoid potential deadlock where SysOut and SysErr almost simultaneously. We'ved moved this out of synchronzation block.
+         *
+         * Context:
+         * Thread 1 > Print of some sort > (locks System Out stream) > .... > (locks FileLogHolder ) Write to file ( no memory causes failed to write to file error) > write to
+         * System err (Thread 2 holds lock)
+         * Thread 2 > System Err ( locks System Err stream) > ... > Write to file (Thread 1 holds FileLogHolder locks)
+         *
+         * This message may be ba bit delayed, but this will avoid deadlocks
+         *
+         */
+        // to avoid junit test to print an error message
+        if (isError.get().booleanValue() == true && System.getProperty("test.classesDir") == null && System.getProperty("test.buildDir") == null) {
```

Comment:
Since this is moved out of the synchronization block, where it use to be emitted right when error occurs, it may be delayed to avoid a dead lock situation.